### ISSUE TITLE
Possibility to keep sampled data in memory?!?

### ISSUE THREAD

**Issue by procrash** [deepmind/reverb#56]
Is there a possibility to keep the sampled data in memory to later on sample it again?
Haven't found anything in the docs so therefore I raised the issue....

**Comment by acassirer** [comment]
Hey,

I'm afraid I don't really understand the question. The data is kept alive (in memory) until deleted from the table and it can be sampled any number of times for as long as it exists.

**Comment by procrash** [comment]
I configure currently the server like this:
`        # Start a Reverb server
        server = reverb.Server(tables=[
            reverb.Table(
                name='the_mother_of_all_queues',
                sampler=reverb.selectors.Fifo(),
                remover=reverb.selectors.Fifo(),
                max_size=self.REVERB_QUEUE_SIZE,
                # Sampling is blocked until inserts > min_size_to_sample
                # "error" = number_of_inserts * samples_per_insert - number_of_samples
                #  Insert calls will block if the error is larger than error_buffer[1],
                #  and sampling will block if the error is less than error_buffer[0].
                rate_limiter=reverb.rate_limiters.SampleToInsertRatio(
                    samples_per_insert=1.0, min_size_to_sample=1,
                    error_buffer=(0, self.REVERB_QUEUE_SIZE)),
                max_times_sampled=1
            ),
        ],
            port=self.reverb_port
        )

`

and sample the server with the following client code:
`
import reverb_dataset
import time
#dataset = reverb_dataset.reverb_dataset_test().as_numpy_iterator()

dataset = reverb_dataset.reverb_dataset().as_numpy_iterator()

start = time.time()

avgTime = 0
nrAvg = 0

while True:
    try:
        # get the next item        
        element = next(dataset)
        diff = time.time() - start        
        avgTime += diff
        nrAvg += 1

        print(f'Fetching one batch took {diff} seconds. Avg:{avgTime/nrAvg}')
        start = time.time()
        # do something with element
    except StopIteration:
        print("all elements fetched")
        # if StopIteration is raised, break from loop
        break
`

Is this keeping the loaded data in memory if the REVERB_QUEUE_SIZE is a huge as the number of samples? Or does reverb delete it after it has been sampled? 

What needs to be changed in order not to delete the sampled data?!?

**Comment by acassirer** [comment]
Hey,

That setup will delete the item straight away after it has been sampled. This is due the `max_times_sampled=1` in your config and is the expected behaviour of a queue.

It sounds like you are not really trying to create a queue but maybe a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer)? You can find examples of how to set this and some other common setups in the [demo colab](https://github.com/deepmind/reverb/blob/master/examples/demo.ipynb).

_Sidenote: Whenever you are actually in need of a queue, a simpler way to create it is to use [`reverb.Table.queue`](https://github.com/deepmind/reverb/blob/master/reverb/server.py#L165)._

**Comment by procrash** [comment]
If I'd remove the max_times_sampled simply would that work as well? And would act like a circular buffer or do I have to take the circular buffer statement from your example link?

**Comment by acassirer** [comment]
Sorry, I now realise that my previous comment about a circular buffer doesn't make much sense since that one will also remove the item after one sample. My comment about removing the `max_times_sampled` still stands though and deleting that will keep the samples in memory until you remove them (either by inserting more items that what fits or by explicitly deleting items).

TLDR: `max_times_sampled=0` will keep the sample in memory. No need to follow my demo colab exactly but it is a useful reference.

**Comment by procrash** [comment]
TLDR: max_times_sampled=0 will keep the sample in memory
Perfect, that's what I was looking for...

**Comment by acassirer** [comment]
No problem. Good luck with your research!
